src/main/java/com/artipie/npm/proxy/http/DownloadPackageSlice.java

```diff
@@ -0,0 +1,125 @@
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.StreamSupport;
+import org.cactoos.map.MapEntry;
+import org.reactivestreams.Publisher;
+
+/**
+ * HTTP slice for download package requests.
+ * @since 0.1
+ * @checkstyle ReturnCountCheck (200 lines)
+ */
+public final class DownloadPackageSlice implements Slice {
+    /**
+     * Download package request URL regexp pattern.
+     */
+    public static final Pattern PATH_PATTERN = Pattern.compile("^/(((?!/-/).)+)$");
+
+    /**
+     * NPM Proxy facade.
+     */
+    private final NpmProxy npm;
+
+    /**
+     * Ctor.
+     *
+     * @param npm NPM Proxy facade
+     */
+    public DownloadPackageSlice(final NpmProxy npm) {
+        this.npm = npm;
+    }
+
+    @Override
+    @SuppressWarnings("PMD.OnlyOneReturn")
+    public Response response(final String line,
+        final Iterable<Map.Entry<String, String>> headers,
+        final Publisher<ByteBuffer> body) {
+        Logger.debug(
+            DownloadPackageSlice.class,
+            "Processing download package request for query %s",
+            line
+        );
+        final Matcher matcher = DownloadPackageSlice.PATH_PATTERN
+            .matcher(new RequestLineFrom(line).uri().getPath());
+        if (!matcher.matches()) {
+            return RsProvider.notFound();
+        }
+        final String name = matcher.group(1);
+        Logger.debug(DownloadPackageSlice.class, "Determined package name is: %s", name);
+        return new AsyncResponse(
+            this.npm.getPackage(name)
+                .map(
+                    pkg -> (Response) new RsWithHeaders(
+                        new RsWithBody(
+                            new RsWithStatus(RsStatus.OK),
+                            new Content.From(
+                                DownloadPackageSlice.clientFormat(pkg.data(), headers).getBytes()
+                            )
+                        ),
+                        Arrays.asList(
+                            new MapEntry<>("Content-Type", "application/json"),
+                            new MapEntry<>("Last-Modified", pkg.lastModified())
+                        )
+                    )
+                ).toSingle(RsProvider.notFound())
+                .to(SingleInterop.get())
+        );
+    }
+
+    /**
+     * Transform internal package format for external clients.
+     * @param data Internal package data
+     * @param headers Request headers
+     * @return External client package
+     */
+    private static String clientFormat(final String data,
+        final Iterable<Map.Entry<String, String>> headers) {
+        final String host = StreamSupport.stream(headers.spliterator(), false)
+            .filter(e -> e.getKey().equalsIgnoreCase("Host"))
+            .findAny().orElseThrow(
+                () -> new RuntimeException("Could not find Host header in request")
```

Comment:
HTTP 500 with error stacktrace

Reply:
@brastak should not it be 400?

Reply:
@olenagerasimova this situation is almost impossible. I use orElseThrow just because of I do not want to use get() call: it can cause warnings in some checkers without isPresent-like checks. But Host header is mandatory in HTTP 1.1, so if it will really absent in some calls, HTTP 500 shows the severity better.

Reply:
@brastak agree, thanks for the explanation